### ISSUE TITLE
Eval disabled from generated functions in 0.5

### ISSUE THREAD

**Issue by andyferris** [MichaelHatherly/AnonymousTypes.jl#1]
Hi Michael!

Great package. I really like the clean approach here. I was hacking over at [GeneratedTypes.jl](https://github.com/andyferris/GeneratedTypes.jl) on a very similar idea.

Unfortunately, `eval` was disabled from `@generated` functions in 0.5 since JuliaLang/julia#16040. I'm still wondering if there is a safe way around this - have you thought about this very much? I heard that "Oscar has a branch" (a phrase I heard repeated many times at JuliaCon!) that implements structs just like this package, but we won't see that until 0.6, and it would be great to have something like this in the mean-time!

Andy

**Comment by MichaelHatherly** [comment]
Thanks, yeah I've been meaning to take a look at this again, but other things have gotten in the way.

I see two possible approaches that could be taken here:
- Try move `eval` one level up into the macro stage, which may be possible here since the number of fields is static. Don't think that could work for your package though since it can do arbitrary(?) computation at generation-time.
- Attempt to use the same approach taken by `Base.deserialize`, via `ccall`ing the right things, when it regenerates types from file. Not sure whether that would technically get around the `eval` though.

**Comment by andyferris** [comment]
Hmm I suppose it _might_ work to only need the field names, and not the types (since the macro wouldn't always be able to determine them). That would be really neat!

For the fully-generated case, I was thinking a `ccall` or two would do the trick... but it just felt so _evil_... :)

**Comment by MichaelHatherly** [comment]
> only need the field names, and not the types

Each generated type would just need to be parametrised over all it's fields like so:

``` julia
immutable GENERATED_NAME{T_x, T_y, T_z}
    x :: T_x
    y :: T_y
    z :: T_z
end
```

and that should do the trick I believe.

To get the dispatch on fieldname that this package currently provides would probably just need some extra type parameters to describe it properly.

> but it just felt so evil... :)

Yeah, pretty much how I was thinking too. I think that's likely to run into the same trouble as using `eval` would.

---

Another approach to dealing with the fully-generated case could be to move the code-gen into the field access itself. What I mean by that is something like the following:

https://gist.github.com/MichaelHatherly/9a86c1bbc661230a3fb28a0039f0f9ed

Which can be use like so:

``` julia
julia> import VirtualFields: @~, @struct

julia> a = @struct x = 1 y = (1,2,3)
VirtualFields.GenType{<x,y>,Tuple{Int64,Tuple{Int64,Int64,Int64}}}((1,(1,2,3)))

julia> a~x
1

julia> a~y
(1,2,3)
```

If we ever get `getfield` overloading built into the language then the `@~` can just be replaced with `.`. `VirtualFields` only does basic field access, but I believe there's enough information there to do more complex things like what your `GeneratedTypes` does as well.

**Comment by MichaelHatherly** [comment]
Having made some adjustments/simplifications to the linked gist, it seems that it's actually relatively easy to generate custom types with this approach. For example, at the end of the file I've defined an `ArbitraryLength{S, N}` type "template" that has `N` fields with field names prefixed with the symbol `S`. A pretty trivial example, but you can then create new instances of it with:

``` julia
julia> t = GenType(ArbitraryLength{:x_, 5}, (0,1,2,3,4));

julia> t~x_5
4
```

and all that's required is a `fields` method that returns the vector of field names. `fields` has access to the type parameters so you can encode pretty much anything in it to determine the names of the returned fields.

**Comment by MichaelHatherly** [comment]
And as far as I can tell the indirection from storing the fields in the `.data` field as a `Tuple` rather that actual fields isn't really noticeable.

**Comment by andyferris** [comment]
Yes, all of that is fine, I believe this is a bit like the approach taken in _TypedTables_, using `getindex` with a `Val`, (with the additional of using a macro as a helper).

Now I'm confused a bit by `~` - where is the `@`? Is this because `~` is an operator?

**Comment by MichaelHatherly** [comment]
The `~` is the special-cased macro that looks like an operator. It's probably going to be deprecated and removed at some point, but currently it looks as though `getfield` overloading is coming in `0.6` so that's fine.

**Comment by andyferris** [comment]
Right, thanks, I did some digging on it today. Interesting operator.
